File: TaskGame-Hero/src/main/java/net/fred/taskgame/hero/fragments/BuyCardsFragment.java
package net.fred.taskgame.hero.fragments;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.fred.taskgame.hero.R;
import net.fred.taskgame.hero.activities.MainActivity;
import net.fred.taskgame.hero.models.Card;
import net.fred.taskgame.hero.models.adapters.BuyCardsAdapter;
import net.fred.taskgame.hero.models.adapters.RecyclerViewItemListener;
import net.fred.taskgame.hero.utils.TaskGameUtils;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;


public class BuyCardsFragment extends BaseFragment {

    @BindView(R.id.recycler_view)
    RecyclerView mRecyclerView;

    private List<Card> mNonObtainedCardList;
    private BuyCardsAdapter mAdapter;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View layout = inflater.inflate(R.layout.fragment_buy_cards, container, false);
        ButterKnife.bind(this, layout);

        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        mRecyclerView.setLayoutManager(layoutManager);

        mNonObtainedCardList = Card.getNonObtainedCardList();
        mAdapter = new BuyCardsAdapter(mNonObtainedCardList, new RecyclerViewItemListener() {
            @Override
            public void onItemClicked(int position) {
                Card card = mNonObtainedCardList.get(position);
                try {
                    startActivityForResult(TaskGameUtils.getRequestPointsActivityIntent(getContext(), card.price), card.id);
                } catch (ActivityNotFoundException e) {
                    // TaskGame application is not installed
                }
            }
        });
        mRecyclerView.setAdapter(mAdapter);

        return layout;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            // the user accepted, let's refresh the DB and UI

            for (int i = 0; i < mNonObtainedCardList.size(); i++) {
                Card card = mNonObtainedCardList.get(i);
                if (card.id == requestCode) { // requestCode is card id
                    getMainActivity().playSound(MainActivity.SOUND_NEW_CARD);
                    card.isObtained = true;
                    card.save();
                    mNonObtainedCardList.remove(i);
                    mAdapter.notifyItemRemoved(i);
                    break;
                }
            }
        }

        super.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    protected int getMainMusicResId() {
        return R.raw.main_theme;
    }
}

File: TaskGame-Hero/src/main/java/net/fred/taskgame/hero/fragments/BuyCardsFragment.kt
/*
 * Copyright (c) 2012-2017 Frederic Julian
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http:></http:>//www.gnu.org/licenses/>.
 */

package net.fred.taskgame.hero.fragments

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.os.Bundle
import android.support.annotation.RawRes
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_buy_cards.*
import net.fred.taskgame.hero.R
import net.fred.taskgame.hero.activities.MainActivity
import net.fred.taskgame.hero.adapters.BuyCardsAdapter
import net.fred.taskgame.hero.adapters.RecyclerViewItemListener
import net.fred.taskgame.hero.models.Card
import net.fred.taskgame.hero.utils.TaskGameUtils
import net.frju.androidquery.gen.CARD
import org.jetbrains.anko.onClick


class BuyCardsFragment : BaseFragment() {

    private var nonObtainedCardList: MutableList<Card>? = null
    private var adapter: BuyCardsAdapter? = null

    override val mainMusicResId: Int
        @RawRes
        get() = R.raw.main_theme

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_buy_cards, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recycler_view.layoutManager = layoutManager
        recycler_view.setEmptyView(empty_view)

        nonObtainedCardList = Card.nonObtainedCardList
        nonObtainedCardList?.let {
            adapter = BuyCardsAdapter(it, object : RecyclerViewItemListener {
                override fun onItemClicked(position: Int) {
                    val card = it[position]
                    if (card.price == 0) {
                        buyCard(position, card)
                    } else {
                        try {
                            startActivityForResult(TaskGameUtils.getRequestPointsActivityIntent(context, card.price.toLong()), card.id)
                        } catch (e: ActivityNotFoundException) {
                            // TaskGame application is not installed
                        }

                    }
                }
            })
        }
        recycler_view.adapter = adapter

        back.onClick { fragmentManager.popBackStack() }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            // the user accepted, let's refresh the DB and UI

            nonObtainedCardList?.let {
                for (i in it.indices) {
                    val card = it[i]
                    if (card.id == requestCode) { // requestCode is card id
                        buyCard(i, card)
                        break
                    }
                }
            }
        }

        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun buyCard(position: Int, card: Card) {
        mainActivity?.playSound(MainActivity.SOUND_NEW_CARD)
        card.isObtained = true
        CARD.save(card).query()
        nonObtainedCardList?.removeAt(position)
        adapter?.notifyItemRemoved(position)
    }
}
